general/forall.hpp

```diff
@@ -780,6 +780,51 @@ inline void forall_3D_grid(int N, int X, int Y, int Z, int G, lambda &&body)
    ForallWrap<3>(true, N, body, X, Y, Z, G);
 }
 
+#ifdef MFEM_USE_MPI
+
+// Function mfem::hypre_forall_gpu() similar to mfem::forall, but it always
+// executes on the GPU device that hypre was configured with at build time.
+#if defined(HYPRE_USING_GPU)
+template<typename lambda>
+inline void hypre_forall_gpu(int N, lambda &&body)
+{
+#if defined(HYPRE_USING_CUDA)
+   CuWrap1D(N, body);
+#elif defined(HYPRE_USING_HIP)
+   HipWrap1D(N, body);
+#else
+#error Unknown HYPRE GPU backend!
+#endif
+}
+#endif
+
+// Function mfem::hypre_forall() similar to mfem::forall, but it executes on the
+// device, CPU or GPU, that hypre was configured with at build time (when the
+// HYPRE version is < 2.31.0) or at runtime (when HYPRE was configured with GPU
+// support at build time and HYPRE's version is >= 2.31.0). This selection is
+// generally independent of what device was selected in MFEM's runtime
+// configuration.
+template<typename lambda>
+inline void hypre_forall(int N, lambda &&body)
+{
+#if !defined(HYPRE_USING_GPU)
+   for (int i = 0; i < N; i++) { body(i); }
```

Comment:
In Hypre, if running on CPU and built with OpenMP, this loop as well as the one below would have the option to be parallelized with threads. Should we do that here too? That is, something like:

```
#if !defined(HYPRE_USING_GPU)
#ifdef MFEM_USE_OPENMP
   OmpWrap(N, body);
#else
   for (int i = 0; i < N; i++) { body(i); }
#endif
...
```

or 

```
#if !defined(HYPRE_USING_GPU)
#ifdef MFEM_USE_OPENMP
   #pragma omp parallel for
#endif
   for (int i = 0; i < N; i++) { body(i); }
...
```

I realize this isn't a common use case but it can have substantial performance impact for users expecting that big loops on device would be parallelized with OpenMP on CPU.

Reply:
We can do that if you want. It makes sense to use `HYPRE_USING_OPENMP` to be consistent with the hypre configuration. Do you know if there's a runtime configuration option in hypre for running sequential/OpenMP on CPU, or is this just the build-time option with `HYPRE_USING_OPENMP`?

Reply:
Nice, I like that idea. I believe it is all build time, everything parallel for loop for host code is guarded as:

```
#ifdef HYPRE_USING_OPENMP
      #pragma omp parallel for HYPRE_SMP_SCHEDULE
#endif
```

Reply:
I did not see any runtime switches in hypre code for OpenMP, so I just pushed a change to use OpenMP when HYPRE_USING_OPENMP is defined -- this is how it is done everywhere in hypre. See b06168ff0.

Reply:
It looks like `HYPRE_SMP_SCHEDULE` is only available in internal headers and only works because `hypre.hpp` and `hypre_parcsr.hpp` include `_hypre_parcsr_mv.h` and `_hypre_parcsr_ls.h`. I get compiler warnings for `forall.hpp` when building which look like:

```
warning: extra tokens at the end of '#pragma omp parallel for' are ignored [-Wextra-tokens]
```

These go away when adding 

```
#ifdef MFEM_USE_MPI
#include <_hypre_utilities.h>
#endif
```

to the top of `forall.hpp`. An alternative would be to just include `hypre.hpp` in `forall.hpp` and rely on the transitive include.

A side note which I believe should also be fixed is that the Hypre header includes at the top of MFEM's `hypre.hpp` and `hypre_parcsr.hpp` are incorrectly using quotation `""` instead of angle bracket `<>` delineations for external library headers.

Reply:
For clarity, here is my full diff:

```
diff --git a/general/forall.hpp b/general/forall.hpp
index 0dc4a75d5..4a397b31b 100644
--- a/general/forall.hpp
+++ b/general/forall.hpp
@@ -19,6 +19,9 @@
 #include "device.hpp"
 #include "mem_manager.hpp"
 #include "../linalg/dtensor.hpp"
+#ifdef MFEM_USE_MPI
+#include <_hypre_utilities.h>
+#endif

 namespace mfem
 {
diff --git a/linalg/hypre.hpp b/linalg/hypre.hpp
index a9e2c4e3d..f6ba0a02b 100644
--- a/linalg/hypre.hpp
+++ b/linalg/hypre.hpp
@@ -16,17 +16,18 @@

 #ifdef MFEM_USE_MPI

+#include "../general/globals.hpp"
+
 #include <mpi.h>

 // Enable internal hypre timing routines
 #define HYPRE_TIMING

 // hypre header files
-#include "seq_mv.h"
-#include "_hypre_parcsr_mv.h"
-#include "_hypre_parcsr_ls.h"
-#include "temp_multivector.h"
-#include "../general/globals.hpp"
+#include <seq_mv.h>
+#include <temp_multivector.h>
+#include <_hypre_parcsr_mv.h>
+#include <_hypre_parcsr_ls.h>

 #ifdef HYPRE_COMPLEX
 #error "MFEM does not work with HYPRE's complex numbers support"
diff --git a/linalg/hypre_parcsr.hpp b/linalg/hypre_parcsr.hpp
index 30938b78d..d6c1063fd 100644
--- a/linalg/hypre_parcsr.hpp
+++ b/linalg/hypre_parcsr.hpp
@@ -16,11 +16,13 @@

 #ifdef MFEM_USE_MPI

+#include "../general/mem_manager.hpp"
+
 // Enable internal hypre timing routines
 #define HYPRE_TIMING

-#include "../general/mem_manager.hpp"
-#include "_hypre_parcsr_mv.h"
+// hypre header files
+#include <_hypre_parcsr_mv.h>

 // Older hypre versions do not define HYPRE_BigInt and HYPRE_MPI_BIG_INT, so we
 // define them here for backward compatibility.
```

Reply:
These look good to me. Do you mind just pushing the changes?

Reply:
Thanks! Done in [73d76bf](https://github.com/mfem/mfem/pull/4279/commits/73d76bf51a4afa02398e111ad2f54684c492c113) and [35e2b1f](https://github.com/mfem/mfem/pull/4279/commits/35e2b1f60ff5d87a7c43586cf50c8f9ce0088ce3).